src/rgw/motr/gc/gc.cc

```diff
@@ -43,51 +45,78 @@ void *MotrGC::GCWorker::entry() {
-        // rc = dequeue();
-
-        // remove the entry from iname
+      ldpp_dout(dpp, 10) << gc_log_prefix
+                      << __func__ << ": Working on GC Queue: " << iname << dendl;
+      
+      bufferlist bl;
+      std::vector<std::string> keys(motr_gc->max_count + 1);
+      std::vector<bufferlist> vals(motr_gc->max_count + 1);
+      uint32_t rgw_gc_obj_min_wait = cct->_conf->rgw_gc_obj_min_wait;
+      keys[0] = obj_exp_time_prefix;
+      rc = motr_gc->store->next_query_by_name(iname,
+                                  keys, vals, obj_exp_time_prefix);
+      ldpp_dout(dpp, 0) << gc_log_prefix 
+                         << __func__ <<": next_query_by_name() rc=" << rc << dendl;
+      if (rc < 0) {
+        // In case of failure, worker will keep retrying till end_time
+        ldpp_dout(dpp, 0) << gc_log_prefix 
+                        << __func__ <<": ERROR: NEXT query failed. rc=" << rc << dendl;
+        continue;
+      }
 
-        // Exit the loop if required work is complete
+      // fetch entries as per defined in rgw_gc_max_trim_chunk from index iname
+      for (uint32_t j = 0; j < motr_gc->max_count &&
+                              !keys[j].empty() && keys[j] != obj_exp_time_prefix; j++) {
+        bufferlist::const_iterator blitr = vals[j].cbegin();
+        motr_gc_obj_info ginfo;
+        ginfo.decode(blitr);
+        // Check if the object is ready for deletion
+        if(ginfo.deletion_time + rgw_gc_obj_min_wait > std::time(nullptr)) {
+          // No more expired object for deletion
+          break;
+        }
+        // delete motr object
+        if(ginfo.is_multipart) {
+          // handle multipart object deletion
+        }
+        else {
+          // simple object
+          rc = motr_gc->delete_motr_obj_from_gc(ginfo);
+          if (rc < 0) {
+            ldpp_dout(dpp, 0) << gc_log_prefix
+                << __func__ <<"ERROR: Motr obj deletion failed for "
+                << ginfo.tag << " with rc: " << rc << dendl;
+            continue; // should continue deletion for next objects
+          }
+        }
+        // delete entry from GC queue
+        rc = motr_gc->dequeue(iname, ginfo);
         processed_count++;
-        if (processed_count >= motr_gc->max_count) break;
+        
         // Update current time
         current_time = std::time(nullptr);
-      } while (current_time < end_time && !motr_gc->going_down());
+        // Exit the loop if required work is complete
+        if (processed_count >= motr_gc->max_count 
+                              || current_time > end_time || motr_gc->going_down())
+          break;
+      }
       // unlock the GC queue
     }
     my_index = (my_index + 1) % motr_gc->max_indices;
 
     // sleep for remaining duration
     // if (end_time > current_time) sleep(end_time - current_time);
-    cv.wait_for(lk, std::chrono::milliseconds(gc_interval * 10));
+    cv.wait_for(lk, std::chrono::seconds(rgw_gc_processor_period));
```

Comment:
We need to wait for the time we did not utilise I think.
```suggestion
    uint32_t unutilized_time = rgw_gc_processor_period) - (end_time - start_time);
    if (unutilize_time > 0)
      cv.wait_for(lk, std::chrono::seconds(unutilized_time);
```

Reply:
Addressed.